flink-tests/src/test/java/org/apache/flink/test/streaming/api/datastream/FinishedSourcesWatermarkITCase.java

```diff
@@ -0,0 +1,131 @@
+public class FinishedSourcesWatermarkITCase extends TestLogger {
+
+    private static final AtomicLong ACTUAL_DOWNSTREAM_WATERMARK = new AtomicLong(0);
+
+    @Test
+    public void testTwoConsecutiveFinishedTasksShouldPropagateMaxWatermark() throws Exception {
+        Configuration conf = new Configuration();
+        StreamExecutionEnvironment env = StreamExecutionEnvironment.createLocalEnvironment(1);
+
+        // disable chaining to make sure we will have two consecutive checkpoints with Task ==
+        // FINISHED
+        env.disableOperatorChaining();
+        env.enableCheckpointing(100);
+
+        // create our sources - one that will want to run forever, and another that finishes
+        // immediately
+        DataStream<String> runningStreamIn =
+                env.addSource(new LongRunningSource(), "Long Running Source");
+        DataStream<String> emptyStream =
+                env.addSource(new ShortLivedEmptySource(), "Short Lived Source");
+
+        // pass the empty stream through a simple map() function
+        DataStream<String> mappedEmptyStream = emptyStream.map(v -> v).name("Empty Stream Map");
+
+        // join the two streams together to see what watermark is reached during startup and after a
+        // recovery
+        runningStreamIn
+                .connect(mappedEmptyStream)
+                .process(new MyCoProcessFunction())
+                .name("Join")
+                .addSink(
+                        new SinkFunction<String>() {
+                            @Override
+                            public void writeWatermark(
+                                    org.apache.flink.api.common.eventtime.Watermark watermark) {
+                                ACTUAL_DOWNSTREAM_WATERMARK.set(watermark.getTimestamp());
+                            }
+                        });
+
+        env.execute();
+    }
+
+    private static class LongRunningSource extends RichSourceFunction<String>
+            implements CheckpointListener {
+        private volatile boolean isRunning = true;
+        private transient long lastSuccessfulCheckpointId;
+
+        @Override
+        public void run(SourceContext<String> sourceContext) throws Exception {
+            long expectedWatermark = Watermark.MAX_WATERMARK.getTimestamp();
+            long watermark = DateTime.now().getMillis();
+            sourceContext.emitWatermark(new Watermark(watermark));
+
+            while (isRunning && expectedWatermark > ACTUAL_DOWNSTREAM_WATERMARK.get()) {
+                synchronized (sourceContext.getCheckpointLock()) {
+                    watermark = DateTime.now().getMillis();
+                    sourceContext.emitWatermark(new Watermark(watermark));
+                    if (lastSuccessfulCheckpointId == 5) {
+                        throw new RuntimeException("Force recovery");
+                    }
+                    if (lastSuccessfulCheckpointId > 10 && expectedWatermark > watermark) {
```

Comment:
```suggestion
                    if (lastSuccessfulCheckpointId == 10) {
```
The latter condition will usually only evaluate to true the first time it is called (excluding cases where `DateTime.now()` goes back in time).

Reply:
Re this and above. This is kind of the point here. We are waiting until 10th checkpoint. On the 10th checkpoint, we are recoding what was the last emitted watermark, and remembering it to be the `expectedWatermark`. Then we are still running unknown number of iterations, until the downstream sink finally receives this `expectedWatermark`. Hence we can not get rid of `expectedWatermark > ACTUAL_DOWNSTREAM_WATERMARK.get()`.

I will try to rephrase the code a bit, to make it more explicit/clear.